File: displayiq/enhancement.py
import cv2
import numpy as np

def enhance_contrast(img: np.ndarray, clip_limit=2.0, tile_grid_size=(8, 8)) -> np.ndarray:
    """Applies CLAHE (Contrast Limited Adaptive Histogram Equalization)."""
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    clahe = cv2.createCLAHE(clipLimit=clip_limit, tileGridSize=tile_grid_size)
    enhanced = clahe.apply(gray)
    return cv2.cvtColor(enhanced, cv2.COLOR_GRAY2BGR)

File: cpp/src/enhancement.cpp
#include "displayiq/enhancement.hpp"

namespace displayiq {

    cv::Mat ImageEnhancer::enhanceContrast(const cv::Mat& inputImage,
        double clipLimit,
        cv::Size tileGridSize) {
        if (inputImage.empty()) {
            throw std::invalid_argument("Input image is empty.");
        }

        cv::Mat gray;
        cv::cvtColor(inputImage, gray, cv::COLOR_BGR2GRAY);

        cv::Ptr<cv::CLAHE> clahe = cv::createCLAHE(clipLimit, tileGridSize);
        cv::Mat enhancedGray;
        clahe->apply(gray, enhancedGray);

        cv::Mat result;
        cv::cvtColor(enhancedGray, result, cv::COLOR_GRAY2BGR);
        return result;
    }

} // namespace displayiq